platform/security/idp/security-idp-server/src/main/java/org/codice/ddf/security/idp/server/IdpEndpoint.java

```diff
@@ -232,6 +262,178 @@ public void init() {
         new ResponseUnmarshaller());
   }
 
+  @Override
+  public Map<String, Set<String>> getActiveSessions() {
+    try {
+      return Security.getInstance().runWithSubjectOrElevate(this::getActiveSessionsSecure);
+    } catch (SecurityServiceException | InvocationTargetException e) {
+      SecurityLogger.audit("Failed to run command; insufficient permissions");
+    }
+
+    return Collections.emptyMap();
+  }
+
+  @VisibleForTesting
+  Map<String, Set<String>> getActiveSessionsSecure() {
+    return cookieCache.getAllSamlSubjects(securityManager);
+  }
+
+  @Override
+  public void invalidateSession(String subjectName) {
+    try {
+      Security.getInstance()
+          .runWithSubjectOrElevate(
+              (Callable<Void>)
+                  () -> {
+                    invalidateSessionSecure(subjectName);
+                    return null;
+                  });
+    } catch (SecurityServiceException | InvocationTargetException e) {
+      SecurityLogger.audit("Failed to run command; insufficient permissions");
+    }
+  }
+
+  @VisibleForTesting
+  void invalidateSessionSecure(String subjectName) {
+    String cacheKey = cookieCache.getCacheKeyBySubjectName(subjectName, securityManager);
+    if (cacheKey == null) {
+      LOGGER.debug(
+          "No cache element found for subject name {}; skipping invalidation.", subjectName);
+      return;
+    }
+
+    LogoutState logoutState = new LogoutState(getActiveSps(cacheKey));
+
+    // TODO: 1/3/18 sufficient?
```

Comment:
@stustison I believe setting this to an empty string - a value it would never be in reality - is sufficient for our purposes. Do you agree?

Reply:
Looking again, I don't think I need to set that field at all. We use it internally for useragent-initiated logouts to ensure we don't send the useragent back to the initial SP to try to logout again, but in our case we want to run through all of the attached SPs once and only once anyway.